File: amic-platform/src/modules/fdd/types/deal.ts
export type DealType = "COMPLETION_ACCOUNTS" | "LOCKED_BOX";
export type DealStatus = "DRAFT" | "ACTIVE" | "ARCHIVED";
export type DefinitionStatus = "DRAFT" | "APPROVED" | "LOCKED";
export type SnapshotStatus = "RUNNING" | "SUCCESS" | "FAILED";
export type DealPhase = "MOU" | "VDR_SETUP" | "DATA_UPLOAD" | "ANALYSIS" | "REPORTING";

import type { IndustryId } from "@/types/industry";
/** @deprecated Use IndustryId from @/types/industry directly */
export type IndustryType = IndustryId;

export interface Deal {
  id: string;
  name: string;
  deal_type: DealType;
  base_currency: string;
  reference_date: string;
  period_start: string;
  period_end: string;
  status: DealStatus;
  created_by: string;
  created_at: string;
  updated_at: string;
  client_name: string | null;
  client_contact_name: string | null;
  client_contact_email: string | null;
  target_company_name: string | null;
  team_partner_id: string | null;
  team_manager_id: string | null;
  scope_qoe: boolean;
  scope_nwc: boolean;
  scope_debt: boolean;
  industry: IndustryType;
  current_phase: DealPhase;
}

export interface DealCreate {
  name: string;
  deal_type: DealType;
  base_currency: string;
  reference_date: string;
  period_start: string;
  period_end: string;
  client_name?: string;
  client_contact_name?: string;
  client_contact_email?: string;
  target_company_name?: string;
  team_partner_id?: string;
  team_manager_id?: string;
  scope_qoe?: boolean;
  scope_nwc?: boolean;
  scope_debt?: boolean;
  industry?: IndustryType;
}

export interface DefinitionData {
  cash: { include: string[]; exclude: string[] };
  debt: { include: string[]; exclude: string[] };
  debt_like: Array<Record<string, unknown>>;
  cash_like: Array<Record<string, unknown>>;
  nwc: { include: string[]; exclude: string[] };
  target_nwc: { method: string; value: number | null };
  lease_ifrs16: { include_in_debt: boolean };
}

export interface DealDefinition {
  id: string;
  deal_id: string;
  version: number;
  definition_data: DefinitionData;
  status: DefinitionStatus;
  approved_by: string | null;
  approved_at: string | null;
  hash: string;
  created_at: string;
  updated_at: string;
}

export interface DealSnapshot {
  id: string;
  deal_id: string;
  definition_version_id: string;
  engine_version: string;
  input_hash: string;
  result_hash: string | null;
  status: SnapshotStatus;
  created_at: string;
}

// ── Upload Types ────────────────────────────────────────

export type UploadType = "TB" | "GL" | "AR" | "AP" | "BANK" | "DEBT" | "LEASE";
export type IngestionStatus =
  | "PENDING"
  | "DETECTING"
  | "VALIDATING"
  | "INGESTING"
  | "COMPLETED"
  | "FAILED";
export type ValidationSeverity = "ERROR" | "WARNING";

export interface ValidationSummary {
  rows_ingested?: number;
  total_warnings?: number;
  total_errors?: number;
  [key: string]: unknown;
}

export interface UploadFile {
  id: string;
  deal_id: string;
  original_filename: string;
  file_hash: string;
  file_size_bytes: number;
  detected_type: UploadType | null;
  confirmed_type: UploadType | null;
  detection_confidence: number | null;
  status: IngestionStatus;
  total_rows: number | null;
  rows_processed: number | null;
  sheet_name: string | null;
  error_message: string | null;
  validation_summary: ValidationSummary | null;
  uploaded_by: string;
  created_at: string;
  updated_at: string;
}

export interface ValidationErrorItem {
  id: string;
  severity: ValidationSeverity;
  error_code: string;
  field_name: string | null;
  row_number: number | null;
  message: string;
  suggestion: string | null;
}

export interface UploadFileDetail extends UploadFile {
  validation_errors: ValidationErrorItem[];
}

File: fdd/backend/app/schemas/deal.py
import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.deal import (
    DealPhase,
    DealStatus,
    DealStructure,
    DealType,
    DefinitionStatus,
    IndustryType,
    InvestmentType,
    SellerType,
    SnapshotStatus,
)

# ── Deal ──────────────────────────────────────────────────


class DealCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    target_company_name: str = Field(..., min_length=1, max_length=255)
    deal_type: DealType = DealType.COMPLETION_ACCOUNTS
    base_currency: str = Field(default="KRW", max_length=10)
    reference_date: date | None = None
    period_start: date | None = None
    period_end: date | None = None
    created_by: str = Field(default="system", max_length=100)

    # Workflow fields (optional at creation)
    client_name: str | None = Field(default=None, max_length=255)
    client_contact_name: str | None = Field(default=None, max_length=255)
    client_contact_email: str | None = Field(default=None, max_length=255)
    team_partner_id: uuid.UUID | None = None
    team_manager_id: uuid.UUID | None = None
    scope_qoe: bool = True
    scope_nwc: bool = True
    scope_debt: bool = True
    industry: IndustryType = IndustryType.GENERAL

    # Deal classification
    deal_structure: DealStructure | None = None
    investment_type: InvestmentType | None = None
    seller_type: SellerType | None = None


class DealUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    deal_type: DealType | None = None
    base_currency: str | None = Field(default=None, max_length=10)
    reference_date: date | None = None
    period_start: date | None = None
    period_end: date | None = None
    status: DealStatus | None = None

    # Workflow fields
    client_name: str | None = Field(default=None, max_length=255)
    client_contact_name: str | None = Field(default=None, max_length=255)
    client_contact_email: str | None = Field(default=None, max_length=255)
    target_company_name: str | None = Field(default=None, max_length=255)
    team_partner_id: uuid.UUID | None = None
    team_manager_id: uuid.UUID | None = None
    scope_qoe: bool | None = None
    scope_nwc: bool | None = None
    scope_debt: bool | None = None
    industry: IndustryType | None = None
    current_phase: DealPhase | None = None

    # Deal classification
    deal_structure: DealStructure | None = None
    investment_type: InvestmentType | None = None
    seller_type: SellerType | None = None


class DealRead(BaseModel):
    id: uuid.UUID
    name: str
    deal_type: DealType
    base_currency: str
    reference_date: date | None = None
    period_start: date | None = None
    period_end: date | None = None
    status: DealStatus
    created_by: str
    created_at: datetime
    updated_at: datetime

    # Workflow fields
    client_name: str | None = None
    client_contact_name: str | None = None
    client_contact_email: str | None = None
    target_company_name: str | None = None
    team_partner_id: uuid.UUID | None = None
    team_manager_id: uuid.UUID | None = None
    scope_qoe: bool = True
    scope_nwc: bool = True
    scope_debt: bool = True
    industry: IndustryType = IndustryType.GENERAL
    current_phase: DealPhase = DealPhase.MOU

    # Deal classification
    deal_structure: DealStructure | None = None
    investment_type: InvestmentType | None = None
    seller_type: SellerType | None = None

    model_config = {"from_attributes": True}


# ── Deal Definition ───────────────────────────────────────


class DefinitionData(BaseModel):
    """SPA/LOI 정의를 설정값으로 표현하는 스키마."""

    cash: dict[str, list[str]] = Field(
        default_factory=lambda: {"include": [], "exclude": []},
        description="Cash 정의: 포함/제외 계정 목록",
    )
    debt: dict[str, list[str]] = Field(
        default_factory=lambda: {"include": [], "exclude": []},
        description="Debt 정의: 포함/제외 계정 목록",
    )
    debt_like: list[dict[str, Any]] = Field(
        default_factory=list,
        description="debt-like 항목 목록 [{item, category, rationale}]",
    )
    cash_like: list[dict[str, Any]] = Field(
        default_factory=list,
        description="cash-like 항목 목록",
    )
    nwc: dict[str, list[str]] = Field(
        default_factory=lambda: {"include": [], "exclude": []},
        description="NWC 정의: 포함/제외 계정 목록",
    )
    target_nwc: dict[str, Any] = Field(
        default_factory=lambda: {"method": "6M_AVG", "value": None},
        description="Target NWC(peg) 산정 방법 및 값",
    )
    lease_ifrs16: dict[str, bool] = Field(
        default_factory=lambda: {"include_in_debt": False},
        description="IFRS 16 리스부채 Net Debt 포함 여부",
    )


class DealDefinitionCreate(BaseModel):
    definition_data: DefinitionData


class DealDefinitionRead(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    version: int
    definition_data: dict[str, Any]
    status: DefinitionStatus
    approved_by: str | None
    approved_at: datetime | None
    hash: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DealDefinitionApprove(BaseModel):
    approved_by: str = Field(..., min_length=1, max_length=100)


# ── Deal Snapshot ─────────────────────────────────────────


class DealSnapshotCreate(BaseModel):
    definition_version_id: uuid.UUID


class DealSnapshotRead(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    definition_version_id: uuid.UUID
    engine_version: str
    input_hash: str
    result_hash: str | None
    status: SnapshotStatus
    created_at: datetime

    model_config = {"from_attributes": True}
